File: xplaincv-screening/backend/app/routers/batches.py
from fastapi import APIRouter, Depends, File, Form, HTTPException, Query, UploadFile

from app.config import settings
from app.deps import get_batch_service
from app.schemas.batch import BatchCreateResponse, BatchDetail, BatchSummary
from app.services.batch_service import BatchNotFoundError, BatchService
from app.services.profile_service import ProfileNotBuiltError, ProfileNotFoundError

router = APIRouter(prefix="/api/batches", tags=["batches"])


@router.post("", response_model=BatchCreateResponse, status_code=202)
async def create_batch(
    profile_id: int = Form(...),
    files: list[UploadFile] = File(...),
    service: BatchService = Depends(get_batch_service),
):
    payload = [(f.filename, await f.read()) for f in files]
    try:
        return service.create(profile_id, payload)
    except ProfileNotFoundError as exc:
        raise HTTPException(404, str(exc)) from exc
    except ProfileNotBuiltError as exc:
        raise HTTPException(409, str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(422, str(exc)) from exc


@router.get("", response_model=list[BatchSummary])
def list_batches(service: BatchService = Depends(get_batch_service)):
    return service.list_all()


@router.get("/{batch_id}", response_model=BatchDetail)
def get_batch(
    batch_id: int,
    page: int = Query(1, ge=1),
    page_size: int = Query(settings.RESULTS_PAGE_SIZE, ge=1, le=100),
    service: BatchService = Depends(get_batch_service),
):
    try:
        return service.get_detail(batch_id, page, page_size)
    except BatchNotFoundError as exc:
        raise HTTPException(404, str(exc)) from exc

File: xplaincv-screening/frontend/src/api/batches.ts
import { getJson, postForm } from './client'
import type { BatchCreateResponse, BatchDetail, BatchSummary, ResultDetail } from './types'

export const createBatch = (profileId: number, files: File[]) => {
  const form = new FormData()
  form.append('profile_id', String(profileId))
  files.forEach((f) => form.append('files', f))
  return postForm<BatchCreateResponse>('/api/batches', form)
}

export const listBatches = () => getJson<BatchSummary[]>('/api/batches')

export const getBatch = (id: number, page = 1, pageSize = 24) =>
  getJson<BatchDetail>(`/api/batches/${id}?page=${page}&page_size=${pageSize}`)

export const getResult = (id: number) => getJson<ResultDetail>(`/api/results/${id}`)
